### ISSUE TITLE
CVA6 illegal instruction handler failed

### ISSUE THREAD

**Issue by LuminolT** [keystone-enclave/keystone#429]
Hi everyone!

I'm currently testing the keystone enclave on CVA6 with the Genesys 2 FPGA borad.

However, an error occurred when I'm attempting to boot the system from the SD card. 

```
...
system_opcode_insn: Failed to access CSR 0x3a0 from M-mode
sbi_trap_error: hart0: illegal instruction handler failed (error -1)
...
```

The complete error log is attached below for reference.

After reviewing past issues, I noticed that #374 seems to be similar to the problem. Given my limited experience, I haven't been able to make much progress. 

Thanks a lot!

## Error log

```
Hello World!
init SPI
status: 0x0000000000000025
status: 0x0000000000000025
SPI initialized!
initializing SD...
SD command cmd0         response : 01
SD command cmd55        response : 01
SD command cmd41        response : 01
SD command cmd55        response : 01
SD command cmd41        response : 00
sd initialized!
gpt partition table header:
        signature:      5452415020494645
        revision:       00010000
        size:           0000005C
        crc_header:     48DBA1A4
        reserved:       00000000
        current lba:    0000000000000001
        backup lda:     0000000003B9FFFF
        partition entries lba:          0000000000000002
        number partition entries:       00000080
        size partition entries:         00000080
gpt partition entry 00
        partition type guid:    D5F7127456A1134B81DC867174929325
        partition guid:         B1488ADF5567D0458851DE08FB4C7A91
        first lba:      0000000000000800
        last lba:       0000000000014000
        attributes:     0000000000000000
        name:   000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
gpt partition entry 01
        partition type guid:    AF3DC60F838472478E793D69D8477DE4
        partition guid:         F3A7EFBA10FDE242B90AFF078A4808EA
        first lba:      0000000000100000
        last lba:       0000000003B9FFDE
        attributes:     0000000000000000
        name:   000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
gpt partition entry 02
        partition type guid:    00000000000000000000000000000000
        partition guid:         00000000000000000000000000000000
        first lba:      0000000000000000
        last lba:       0000000000000000
        attributes:     0000000000000000
        name:   000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
gpt partition entry 03
        partition type guid:    00000000000000000000000000000000
        partition guid:         00000000000000000000000000000000
        first lba:      0000000000000000
        last lba:       0000000000000000
        attributes:     0000000000000000
        name:   000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
copying boot image ................................ done!

OpenSBI v1.1
   ____                    _____ ____ _____
  / __ \                  / ____|  _ \_   _|
 | |  | |_ __   ___ _ __ | (___ | |_) || |
 | |  | | '_ \ / _ \ '_ \ \___ \|  _ < | |
 | |__| | |_) |  __/ | | |____) | |_) || |_
  \____/| .__/ \___|_| |_|_____/|____/_____|
        | |
        |_|

[SM] Initializing ... hart [0]
system_opcode_insn: Failed to access CSR 0x3a0 from M-mode
sbi_trap_error: hart0: illegal instruction handler failed (error -1)
sbi_trap_error: hart0: mcause=0x0000000000000002 mtval=0x000000003a0027f3
sbi_trap_error: hart0: mepc=0x00000000800125e8 mstatus=0x8000000a00007800
sbi_trap_error: hart0: ra=0x0000000080005018 sp=0x0000000080059ec0
sbi_trap_error: hart0: gp=0x0000000000000000 tp=0x000000008005a000
sbi_trap_error: hart0: s0=0x0000000080059ed0 s1=0x0000000080052690
sbi_trap_error: hart0: a0=0xffffffff00ffffff a1=0x0000000000000000
sbi_trap_error: hart0: a2=0x0000000080025e88 a3=0x0000000000000000
sbi_trap_error: hart0: a4=0x0000000000000001 a5=0x00000000800125e8
sbi_trap_error: hart0: a6=0x0000000000000007 a7=0x0000000000000000
sbi_trap_error: hart0: s2=0xffffffffffffffff s3=0x0000000000000001
sbi_trap_error: hart0: s4=0x0000000080059edc s5=0x0000000080032000
sbi_trap_error: hart0: s6=0x0000000000000000 s7=0x0000000000000001
sbi_trap_error: hart0: s8=0x0000000000002000 s9=0x0000000000000000
sbi_trap_error: hart0: s10=0x0000000000000000 s11=0x0000000000000000
sbi_trap_error: hart0: t0=0x000000008000b4f4 t1=0xffffffffffff00ff
sbi_trap_error: hart0: t2=0x0000000000000008 t3=0xffffffffff00ffff
sbi_trap_error: hart0: t4=0xffffff00ffffffff t5=0xffff00ffffffffff
sbi_trap_error: hart0: t6=0xff00ffffffffffff
```

**Comment by LuminolT** [comment]
#388 gives a method to solve such problem, but it doesn't work for me.

I've tested the [OpenSBI fix](https://github.com/riscv-software-src/opensbi/commit/ec0559eb315bd11a5ef8865a8a8fa91ab6888250) but the error still occurred. The error log is same as previous one.

**Comment by acaldaya** [comment]
they seems different issues, we saw the illegal instruction when we ran the enclave.
In your case it seems to be during boot.

Are you using vanilla Keystone?

**Comment by LuminolT** [comment]
Thanks for your reply. @acaldaya 

Yes I'm using the vanilla one. I follow the [Running Keystone on CVA6](https://docs.keystone-enclave.org/en/latest/Getting-Started/Running-Keystone-on-CVA6.html#building-keystone) from docs to build the Keystone.

**Comment by LuminolT** [comment]
additional information:

I've tested native [CVA6 SDK](https://github.com/openhwgroup/cva6-sdk) and it could be booted successfully.

**Comment by LuminolT** [comment]
After testing, I found that this is an issue with version of CVA6.

Previously, I was using [Ariane 4.2](https://github.com/openhwgroup/cva6/releases/tag/v4.2.0) packaged from the official CVA6 repository, which was last updated in 2019 and might be outdated. I've checked the version with commit number `797f0a90` (which is said be tested on the Keystone manual), and system could be booted successfully.

I'm going to close this issue, thanks for every commenter.
